source/fe/mapping_fe.cc

```diff
@@ -0,0 +1,1385 @@
+            if (update_flags & update_JxW_values)
+              {
+                JxW_values[i] = boundary_forms[i].norm() * weights[i];
+
+                if (subface_no != numbers::invalid_unsigned_int)
+                  {
+                    const double area_ratio=GeometryInfo<dim>::subface_ratio(
+                                              cell->subface_case(face_no), subface_no);
+                    JxW_values[i] *= area_ratio;
+                  }
+              }
+
+            if (update_flags & update_normal_vectors)
+              normal_vectors[i] = Point<spacedim>(boundary_forms[i] / boundary_forms[i].norm());
+          }
+    }
+
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+unsigned int
+MappingFE<dim,spacedim,DH,VECTOR>::get_degree() const
+{
+  return fe->degree;
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+ComponentMask
+MappingFE<dim,spacedim,DH,VECTOR>::get_fe_mask() const
+{
+  return this->fe_mask;
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+Mapping<dim,spacedim> *
+MappingFE<dim,spacedim,DH,VECTOR>::clone () const
+{
+  return new MappingFE<dim,spacedim,DH,VECTOR>(*this);
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+void
+MappingFE<dim,spacedim,DH,VECTOR>::update_internal_dofs (
+  const typename Triangulation<dim,spacedim>::cell_iterator &cell) const
+{
+  if (euler_dof_handler == 0)
+    {
+      std::cout << "euler_dof_handler is empty!" << std::endl;
+      return;
+    }
+
+  typename DH::cell_iterator dof_cell(*cell, euler_dof_handler);
+  Assert (dof_cell->active() == true, ExcInactiveCell());
+
+  dof_cell->get_dof_indices(dof_indices);
+
+  for (unsigned int i=0; i<local_dofs.size(); ++i)
+    {
+      local_dofs[i] = (*euler_vector)(dof_indices[i]);
+    }
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+void MappingFE<dim,spacedim,DH,VECTOR>::update_euler_vector_using_triangulation
+(VECTOR &vector)
+{
+  if ( fe->has_support_points() )
+    {
+      std::vector<Point<dim> > support_points = fe->get_unit_support_points();
+      typename DH::active_cell_iterator cell;
+      Quadrature<dim> quad(support_points);
+
+      MappingQ<dim,spacedim> map_q(fe->degree);
+      FEValues<dim,spacedim> fe_v(map_q, *fe, quad, update_quadrature_points);
+      std::vector<unsigned int> dofs(fe->dofs_per_cell);
+
+      AssertDimension(fe->dofs_per_cell, support_points.size());
+      Assert(fe->is_primitive(), ExcMessage("FE is not Primitive! This won't work."));
+
+      for (cell = euler_dof_handler->begin_active(); cell != euler_dof_handler->end(); ++cell)
+        {
+          fe_v.reinit(cell);
+          cell->get_dof_indices(dofs);
+          const std::vector<Point<spacedim> > &points = fe_v.get_quadrature_points();
+          for (unsigned int q = 0; q < points.size(); ++q)
+            {
+              unsigned int comp = fe->system_to_component_index(q).first;
+              vector(dofs[q]) = points[q][comp];
```

Comment:
don't you have to translate from `comp` to the system index using the component mask you have?